### DISCUSSION TITLE
Cue and Propagator Networks

### DISCUSSION THREAD

**Discussion by verdverm** [cuelang/cue#408] [General]
@mpvl mentioned Cue and Propagator Networks on slack, moving the discussion here.

---
References on Propagator Networks

- [Propagators: An Introduction (slides) (2017)](https://qfpl.io/share/talks/propagators/slides.pdf)
- [The Art of the Propagator (pdf) (2008)](https://dspace.mit.edu/bitstream/handle/1721.1/44215/MIT-CSAIL-TR-2009-002.pdf?sequence=1&isAllowed=y)
- [Alexey Radul PhD Thesis (pdf) (2009)](http://groups.csail.mit.edu/genesis/papers/radul%202009.pdf)
- [HN Discussion (2015)](https://news.ycombinator.com/item?id=10390477)
- [Haskell Implementation (github)](https://github.com/ekmett/propagators)

**Comment by verdverm** [comment]
Is Cue a Propagator Network, a language to write them in, or both?

Reading through the references, Cue already seems highly influenced and capable of representing several of the examples (if not all?). How do the `cells` and `propagators` map onto Cue?

**Reply by mpvl** [reply]
Cue was not influenced by them at all, but it is based on the same theory, so not surprising there is so much overlap.

**Reply by verdverm** [reply]
Maybe that's why I'm seeing Cue as both a PN and a language where one can write PNs

**Comment by verdverm** [comment]
Can propagator networks be taken to statistics (I think yes?) to the (calculus) limit? (unknown but super interesting to ponder what this means)

If in the limit, could be have some sort of auto-differentiation to optimize propagator networks with gradient-descent / backpropagation (it's in the name, so I'm optimistic)

**Comment by rudolph9** [comment]
@ekmett (creator of the [Haskell Implementation (github)](https://github.com/ekmett/propagators)) has a really good [talk on Propagators](https://confengine.com/functional-conf-2019/proposal/12713/propagators)

I actually had very similar thoughts with regard to propagators and reached out to @ekmett late last year. He sent me the following for more information on propagators:

https://www.youtube.com/watch?v=nY1BCv3xn24&feature=youtu.be is George Wilson's introduction to the topic, which does a better job of motivating the structure than I do.
https://www.youtube.com/watch?v=O3tVctB_VSU is Sussman's take on the topic, which focuses heavily on provenance.
https://dspace.mit.edu/handle/1721.1/54635 is Alexey Radul's thesis. (There is also a tech report.)

Also the work of Chris Mieklejohn on LASP: https://www.youtube.com/watch?v=lhHtiBpDa54
and Lindsey's work on LVish https://users.soe.ucsc.edu/~lkuper/ are both pretty relevant, 
as is anything on CRDTs, 
and the existing CPAIOR research on constraint logic programming.

**Comment by verdverm** [comment]
I'm fascinated by the idea that the authors propose in Art of Prop, of using PNs as a mental construct for thinking about how humans and society. How at multiple scales, humans have this amazing ability to hold opposing ideals in their belief systems, to say on thing and do another, how the mass interconnected partial information calculations map really well onto reality.

I'm only beginning my journey into what these systems mean `(tech | social) | (scale: [...number]) | (past | present | future ) | focalLen: [...number]` (started scatting there in cue ;) (side note, are disjunctions one way to "enumerate" or maybe describe, the entire space of a grid search in N dimensions?)

Well, anyway, I'm glad to meet another person interested in PNs @rudolph9 !

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/408.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
